phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixHAExecutorServiceProvider.java

```diff
@@ -0,0 +1,309 @@
+ * ParallelPhoenixConnections Also provides visibility into the capacity of ExecutorService
+ * {@link PhoenixHAExecutorServiceProvider#hasCapacity(Properties)}
+ */
+public class PhoenixHAExecutorServiceProvider {
+
+    public static final String HA_MAX_POOL_SIZE = "phoenix.ha.max.pool.size";
+    public static final String DEFAULT_HA_MAX_POOL_SIZE = "30";
+    public static final String HA_MAX_QUEUE_SIZE = "phoenix.ha.max.queue.size";
+    public static final String DEFAULT_HA_MAX_QUEUE_SIZE = "300";
+    public static final String HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD =
+            "phoenix.ha.threadpool.queue.backoff.threshold";
+    public static final String DEFAULT_HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD = "0.9";
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(PhoenixHAExecutorServiceProvider.class);
+
+    // Can make configurable if needed
+    private static final int KEEP_ALIVE_TIME_SECONDS = 120;
+
+    private static volatile List<ExecutorService> INSTANCE = null;
+
+    private PhoenixHAExecutorServiceProvider() {
+    }
+
+    public static List<ExecutorService> get(Properties properties) {
+        if (INSTANCE == null) {
+            synchronized (PhoenixHAExecutorServiceProvider.class) {
+                if (INSTANCE == null) {
+                    INSTANCE = initThreadPool(properties);
+                }
+            }
+        }
+        return INSTANCE;
+    }
+
+    @VisibleForTesting
+    static synchronized void resetExecutor() {
+        INSTANCE = null;
+    }
+
+    /**
+     * Checks if the underlying executorServices have sufficient available capacity based on
+     * {@link #HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD}. Monitors the capacity of the blockingqueues
+     * linked with the executor services.
+     * @param properties
+     * @return true if queue is less than {@link #HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD} full
+     */
+    public static List<Boolean> hasCapacity(Properties properties) {
+        if (INSTANCE == null) {
+            return ImmutableList.of(Boolean.TRUE, Boolean.TRUE);
+        }
+        double backoffThreshold =
+                Double.parseDouble(properties.getProperty(HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD,
+                    DEFAULT_HA_THREADPOOL_QUEUE_BACKOFF_THRESHOLD));
+        int i = 0;
+        List<Boolean> executorCapacities = new ArrayList<>();
+        for (ExecutorService executor : INSTANCE) {
+            double queueSize = ((ThreadPoolExecutor) executor).getQueue().size();
+            double queueRemainingCapacity =
+                    ((ThreadPoolExecutor) executor).getQueue().remainingCapacity();
+            double queueCapacity = queueSize + queueRemainingCapacity;
+            boolean hasCapacity = ((queueSize / queueCapacity) < backoffThreshold);
+            if (hasCapacity) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.info("PhoenixHAExecutorServiceProvider ThreadPoolExecutor[" + i
```

Comment:
isDebugEnabled() wrapping an info() call. Also, won't this be very frequent even for a DEBUG?

Reply:
Kept the !hasCapacity block.